[BEAM-314] Add zip compression support in TextIO

jbonofre opened the pull request:
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [X] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [X] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [X] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [X] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---
Add zip compression support in TextIO.

davorbonaci commented:
R: @dhalperi

dhalperi commented on sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java:
```diff
@@ -404,6 +406,37 @@ public void testCompressedRead() throws Exception {
 
   @Test
   @Category(NeedsRunner.class)
+  public void testZipCompressedRead() throws Exception {
```
please add a test with an empty (but valid) zip file.

jbonofre replied:
Good idea, let me do it.

dhalperi commented on sdks/java/core/src/main/java/org/apache/beam/sdk/io/CompressedSource.java:
```diff
@@ -149,6 +150,26 @@ public ReadableByteChannel createDecompressingChannel(ReadableByteChannel channe
+        if (zip.getNextEntry() != null) {
+          return Channels.newChannel(zip);
+        }
+        return null;
```
should you instead throw an `UnsupportedOperationException` explaining the bug?

And maybe file a related JIRA?

dhalperi commented on sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java:
```diff
@@ -404,6 +406,37 @@ public void testCompressedRead() throws Exception {
+    writer.close();
+    out.close();
+
+    Pipeline p = TestPipeline.create();
```
suggest extracting lines 411-426 in a utility function that will create a zip file. Then you can reuse it for multiple zip file tests.

dhalperi replied:
still think it would be a good idea to try to refactor the tests to reduce code reuse, if you can take a stab

dhalperi commented on sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java:
```diff
@@ -404,6 +406,37 @@ public void testCompressedRead() throws Exception {
 
   @Test
   @Category(NeedsRunner.class)
```
(@jbonofre no action needed, this is just copying another test and has no new issues w.r.t. this comment. -- comments below are just a heads-up)

@tgroh I'm sure you're thinking about these tests... Needs runner and also needs local filesystem to be the same system as pipeline. As we improve things, should we be (e.g.) using IOChannelUtils and such instead to create the temp files?

jbonofre replied:
Not sure I understand what you mean. You mean that @Category is not required ?

dhalperi replied:
@jbonofre nope, no changes needed in this PR; not asking anything from you.

For now, just trying to seed a conversation about the `@Category` tag in general. Specifically, "needs runner" is not currently descriptive enough. This test wouldn't run in Cloud Dataflow and probably also not in a Flink or Spark cluster -- it relies on the local file system.

tgroh replied:
Yes. `NeedsRunner` isn't really descriptive enough for what we want. This test requires a capability of Local File System Access.

Alternatively, the test could be written to access the `tempLocation` via `IOChannelUtils`

jbonofre replied:
Thanks for the update guys. I now understand better ;)

dhalperi commented on sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java:
```diff
@@ -404,6 +406,37 @@ public void testCompressedRead() throws Exception {
+    Pipeline p = TestPipeline.create();
+
+    TextIO.Read.Bound<String> read =
+        TextIO.Read.from(filename).withCompressionType(CompressionType.ZIP);
```
also add a test with default mode, which should auto-detect the ZIP files based on extension

jbonofre replied:
Good idea, I'm doing that.

jbonofre commented:
Rebased and updated. I have to figure out the expectException issue in the test.

dhalperi commented on sdks/java/core/src/main/java/org/apache/beam/sdk/io/CompressedSource.java:
```diff
@@ -149,6 +150,26 @@ public ReadableByteChannel createDecompressingChannel(ReadableByteChannel channe
+        if (zip.getNextEntry() != null) {
+          return Channels.newChannel(zip);
+        }
+        throw new IllegalArgumentException("ZIP file doesn't contain any entry");
```
What will the behavior be today on a multi-entry zip? Will it silently produce bad data? Fail in some way?

Please comment, and then also add a test.

dhalperi commented on sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java:
```diff
@@ -404,6 +407,63 @@ public void testCompressedRead() throws Exception {
+
+    expectedException.expect(IllegalArgumentException.class);
+    p.apply(TextIO.Read.from(filename).withCompressionType(CompressionType.ZIP));
+
```
missing `p.run()`?

jbonofre commented:
Rebased and updated.

dhalperi commented on sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java:
```diff
@@ -412,6 +414,95 @@ public void testCompressedRead() throws Exception {
+    p.run();
+
+    // test with auto-detect ZIP based on extension.
+    p = TestPipeline.create();
```
please make this a separate unit test -- that way they can pass or fail independently :)

dhalperi commented on sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java:
```diff
@@ -412,6 +414,95 @@ public void testCompressedRead() throws Exception {
+    expectedException.expect(IllegalArgumentException.class);
+    p.apply(TextIO.Read.from(filename).withCompressionType(CompressionType.ZIP));
+
+    p.run();
```
Can you explain this test better? I would expect to see you looking for some sort of PipelineExecutionException

dhalperi replied:
(@tgroh , suggestions for handling the exception?)

tgroh replied:
We don't wrap exceptions thrown from Source methods in `UserCodeExceptions`, so they won't get rewrapped in `PipelineExecutionExceptions`.

I would put the `expectedException` right before the line that is expected to throw, however.

dhalperi commented on sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java:
```diff
@@ -412,6 +414,95 @@ public void testCompressedRead() throws Exception {
+    PCollection<String> output = p.apply(read);
+
+    PAssert.that(output).containsInAnyOrder(expected);
+    p.run();
```
I'm not sure what this is testing. Since all files are empty, I can't tell whether it read 1 of the files, 2 of the files, 3 of the files, or none of the files.

It would be good if all of the entries were non-empty and different, so that we could tell what the desired behavior was and verify that it worked against the other cases.

dhalperi replied:
Based on your code in the actual zip decompresser, I expect this to fail. But the test itself does not expect failure.

dhalperi commented on sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java:
```diff
@@ -412,6 +414,95 @@ public void testCompressedRead() throws Exception {
+
+  @Test
+  @Category(NeedsRunner.class)
+  public void testZipCompressedReadWithEmptyFile() throws Exception {
```
Can you also add a test where the zip file is valid, has a single entry, and that entry is empty? In other words, a zip of an empty file. Seems like another corner case to handle!

dhalperi commented on sdks/java/core/src/main/java/org/apache/beam/sdk/io/CompressedSource.java:
```diff
@@ -149,6 +150,29 @@ public ReadableByteChannel createDecompressingChannel(ReadableByteChannel channe
+
+    /**
+     * Reads a byte channel assuming it is compressed with zip.
+     */
```
can you add more to the Javadoc here?

```
Only zip files that have a single entry are supported.
```

dhalperi commented on sdks/java/core/src/main/java/org/apache/beam/sdk/io/CompressedSource.java:
```diff
@@ -149,6 +150,29 @@ public ReadableByteChannel createDecompressingChannel(ReadableByteChannel channe
+          if (zip.getNextEntry() != null) {
+            throw new IllegalArgumentException("Multi-entries ZIP file is not supported");
+          } else {
+            return Channels.newChannel(zip);
```
does this work? Based on how you wrote the code, I would expect that `zip.getNextEntry()` actually mutates the ZipInputStream in some way. Is it still valid to open a stream from the zip after calling getNextEntry twice?

dhalperi commented:
Hi JB,

This is looking pretty good!

But I have some questions about the tests. Specifically, since we mostly test empty files it seems tough to validate that the decompressor does exactly what we expect.

I've added some suggestions for improvements.

Thanks,
Dan

jbonofre commented:
Updated. I added explanations on each test.
However, I have two points:
1. Using `ZipInputStream` there's not way to get the number of entries. So, unfortunately, I don't see a good way to raise an exception when the zip stream contains multiple entries.
2. In the `testZipCompressedReadWithEmptyFile` test, I got the right `IllegalArgumentException` but wrapped in a `RuntimeException`. That's why I'm doing `expectedException.expect(RuntimeException.class)` instead of `expectedException.expect(IllegalArgumentException.class)`. Any idea why ?

dhalperi commented on sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java:
```diff
@@ -410,6 +412,158 @@ public void testCompressedRead() throws Exception {
+    PCollection<String> output = p.apply(read);
+
+    PAssert.that(output).empty();
+    p.run();
```
I'd really like to find a way that this case either concats all the files or throws an exception. The current behavior is effectively silent data loss, the worst possible case!

One way to handle this could be a utility InputStream class that wraps `ZipInputStream` and under-the-hood concats all the different entry streams. This is probably the base case.

A second possibility is that once the input stream hits EOF, we check for a next entry and only then throw an exception. But this is less desirable as we don't fail until pretty late.

Can you look into it?

jbonofre replied:
I like your idea of wrapping the stream. Let me figure it out.

jbonofre commented:
I updated the PR to use another approach: I'm using directly the `ZipInputStream` with considering the entries. It allows user to read multi-entries zip file as single entry one. I also updated the tests to show the "new" behavior.

dhalperi commented on sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java:
```diff
@@ -410,6 +412,170 @@ public void testCompressedRead() throws Exception {
+
+  /**
+   * Read a ZIP compressed file with multiple entries. Only the first entry is actually read.
+   * We expect an empty PCollection.
```
update the javadoc to be correct.

dhalperi commented on sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java:
```diff
@@ -410,6 +412,170 @@ public void testCompressedRead() throws Exception {
+  }
+
+  /**
+   * Read a ZIP compressed empty file. We expect an exception.
```
update the javadoc?

dhalperi commented on sdks/java/core/src/main/java/org/apache/beam/sdk/io/CompressedSource.java:
```diff
@@ -149,6 +150,25 @@ public ReadableByteChannel createDecompressingChannel(ReadableByteChannel channe
+    /**
+     * Reads a byte channel assuming it is compressed with zip.
+     *
+     * Warning: only zip files that have exactly one entry (aka are ZIPs of a single file) are
```
update the javadoc?

dhalperi commented on sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java:
```diff
@@ -410,6 +412,170 @@ public void testCompressedRead() throws Exception {
+
+    ZipOutputStream out = new ZipOutputStream(new FileOutputStream(tmpFile));
+
+    PrintStream writer = new PrintStream(out);
```
I think you need to close writer before you call `out.closeEntry`, and reopen a new `writer` for every new entry.

dhalperi commented:
Thanks JB -- looking great.

* Please ensure that the multi-entry test is actually passing by fixing the way the writers are used.
* Fix the javadoc with new semantics.

Other than that, I'd love to see less code reuse in the tests. This is considered a strong recommendation, but I won't withhold and LGTM on that basis ;).

jbonofre commented:
My bad about the javadoc. I will fix that. I'm also fixing the multi-entries test. Then I will refactore a bit to use a share method for zip file creation ;)

jbonofre commented:
PR rebased and updated based on Dan's comments.

dhalperi commented on sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java:
```diff
@@ -410,6 +412,148 @@ public void testCompressedRead() throws Exception {
+    String tmpFileName = tmpFile.getPath();
+
+    ZipOutputStream out = new ZipOutputStream(new FileOutputStream(tmpFile));
+    PrintStream writer = new PrintStream(out);
```
Please use

```java
PrintStream writer = new PrintStream(out, true /* auto-flush on write */);
```

this way we can be sure that the `PrintStream` itself does not buffer any data. See [javadoc](https://docs.oracle.com/javase/7/docs/api/java/io/PrintStream.html#PrintStream(java.io.OutputStream,%20boolean)),

dhalperi commented on sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java:
```diff
@@ -410,6 +412,148 @@ public void testCompressedRead() throws Exception {
+   * @return The zip filename.
+   * @throws Exception In case of a failure during zip file creation.
+   */
+  private String createZipFile(List<String> expected, String filename, String[] ... fieldsEntries)
```
Mark any nullable parameters as `@Nullable`

dhalperi commented on sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java:
```diff
@@ -410,6 +412,148 @@ public void testCompressedRead() throws Exception {
+  @Test
+  @Category(NeedsRunner.class)
+  public void testZipCompressedReadWithEmptyFile() throws Exception {
+    String filename = createZipFile(null, null);
```
believe the first parameter here and below should be `new ArrayList<>()` -- that parameter is not `@Nullable` (even though because there is no data in the file it does not crash.

dhalperi commented:
Looking great! Only trivial changes left.

jbonofre commented:
Rebased and updated according to Dan's comments.

jbonofre commented:
There are tests failure in the DirectRunner due to the latest changes and rebase. I will fix that.

dhalperi commented:
Hi JB!

One small request:
* Please don't rebase every time. It's much easier to review if you simply add new incremental CLs to the existing pull request, as this way I can see what changed ;) (This is mentioned briefly in the [contribution guide](http://beam.incubator.apache.org/contribution-guide/#code-review-and-revision), but is not obvious and may not be standard practice everywhere).

I checked out your PR and tried to get the tests to pass -- but I was never able to make both the single-file and multi-file tests pass. I think that right now the ZipInputStream does not automatically concat all file contents without more effort. Might need to make a new class that wraps the ZipInputStream and manually calls `getNextEntry` whenever the current entry reaches EOF.

jbonofre commented:
Hi Dan,

Thanks for the update and ok for the rebase (sorry about that). I'm checking and fixing the issue.

jbonofre commented:
Working on a ZipInputStream wrapper (which call getNextEntry() behind the hood). Update soon.

jbonofre commented:
@dhalperi It should be OK now.

dhalperi commented on sdks/java/core/src/main/java/org/apache/beam/sdk/io/CompressedSource.java:
```diff
@@ -164,12 +166,40 @@ public boolean matches(String fileName) {
+        currentEntry = getNextEntry();
+      }
+
+      public int read(byte[] b, int off, int len) throws IOException {
```
there are a lot of variants of `read` in an `InputStream`. Is it obvious that this is the only variant you need to override? (May well be, I just don't know).

jbonofre replied:
Yes, it's the only read using in the IOChannel.

dhalperi replied:
I do think you should override all applicable functions from InputStream -- implementation of IOChannel might change and/or this stream might be used in a different way.

lukecwik replied:
This is the only read that has to be implemented, all the others are just wrappers for this read method.

int read() is the only one that you may want to handle specially since its very inefficient if its not implemented but its really inefficient for people to use so having an efficient implementation is useful.

dhalperi commented on sdks/java/core/src/main/java/org/apache/beam/sdk/io/CompressedSource.java:
```diff
@@ -149,9 +152,54 @@ public ReadableByteChannel createDecompressingChannel(ReadableByteChannel channe
+            return -1;
+          }
+          result = super.read(b, off, len);
+          return result;
```
what if there are two empty entries in a row?

lukecwik replied:
Even if there is only one entry and its empty, super.read will return -1 representing the end of that entry which will be returned to the caller and that will prematurely end the stream ignoring any remaining zip entries.

jbonofre replied:
I tested the following cases:
1. zip with two entries (non empty)
2. zip with the first entry empty and the second one with data
3. zip with the first entry with data and the second one empty

It works fine:

```
    String filename = createZipFile(new ArrayList<String>(), null, new String[]{}, new
        String[]{"three", "four"});

    FullZipInputStream zipInputStream = new FullZipInputStream(new FileInputStream(new File
        (filename)));
    BufferedReader reader = new BufferedReader(new InputStreamReader(zipInputStream));
    String line;
    while ((line = reader.readLine()) != null) {
      System.out.println(line);
    }
```

dhalperi replied:
Hi JB, Here's a test that fails that exhibits the problem I'm worried about.

```java
  /**
   * Read a ZIP compressed file containing data, multiple empty entries, and then more data. We
   * expect just the data back.
   */
  @Test
  @Category(NeedsRunner.class)
  public void testZipCompressedReadWithComplexEmptyAndPresentEntries() throws Exception {
    String filename = createZipFile(
        new ArrayList<String>(),
        null,
        new String[] {"cat"},
        new String[] {},
        new String[] {},
        new String[] {"dog"});
    List<String> expected = ImmutableList.of("cat", "dog");

    Pipeline p = TestPipeline.create();

    PCollection<String> output =
        p.apply(TextIO.Read.from(filename).withCompressionType(CompressionType.ZIP));
    PAssert.that(output).containsInAnyOrder(expected);

    p.run();
  }
```

dhalperi replied:
I admit two empty files is relatively unlikely, though totally possible. But this is not just some weird corner case--there are other execution paths in which it is possible for the stream to read exactly to the end of a file but not hit EOF (so `-1` is not returned yet), and then it will act like an empty file for the purposes of the next call into the function.

jbonofre replied:
Catcha, fixing that. Thanks !

lukecwik commented on sdks/java/core/src/main/java/org/apache/beam/sdk/io/CompressedSource.java:
```diff
@@ -149,9 +152,54 @@ public ReadableByteChannel createDecompressingChannel(ReadableByteChannel channe
     /**
+     * Extend of {@link ZipInputStream} to automatically read all entries in the zip.
+     */
+    public class FullZipInputStream extends ZipInputStream {
```
Please extend AbstractInputStream in which you delegate to an internal member variable representing the zip input stream. This prevents the exposure of methods on ZipInputStream.

Also, make this class private and state. Users don't need to have visibility of this implementation.

dhalperi replied:
s/state/static/

jbonofre replied:
Thanks for the update guys. I'm updating accordingly.

jbonofre replied:
AbstractInputStream doesn't exist in the JDK. Which one are you talking about ? Thanks !

lukecwik replied:
I meant to have said InputStream since its the abstract base class and not an interface instead of AbstractInputStream.

lukecwik commented on sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java:
```diff
@@ -419,6 +423,150 @@ public void testCompressedRead() throws Exception {
+   * @return The zip filename.
+   * @throws Exception In case of a failure during zip file creation.
+   */
+  private String createZipFile(List<String> expected, @Nullable String filename, String[]
```
formatting?

dhalperi commented:
JB this is awesome. One small fix to get the multiple-eof-in-a-row case fixed and let's merge it. A big accomplishment!

jbonofre commented:
Great feedback guys ! Much appreciated. Fixing the multi-entries. Thanks !

jbonofre commented:
Fixed the `read()` methods and now extends `InputStream` wrapping the `ZipInputStream` instead of extending it.

dhalperi commented:
LGTM. Jenkins flakes on network errors, but Travis is fine. Merging.
